Run prettier on the remainder of the Repo

robert-j-webb opened the pull request.

robert-j-webb commented:
So this prettier run mangles `scriptsa/build.ts`. I added it to `.prettierignore` to avoid any issues

Original
```ts
const inlineRegexSourcePlugin: Plugin = {
	name: 'inline-regex-source',
	renderChunk(code) {
		const str = new MagicString(code);
		str.replace(
			/\/((?:[^\n\r[\\\/]|\\.|\[(?:[^\n\r\\\]]|\\.)*\])+)\/\s*\.\s*source\b/g,
			(m, source: string) => {
				// escape backslashes
				source = source.replace(/\\(.)|\[(?:\\s\\S|\\S\\s)\]/g, (m, g1: string) => {
					if (g1) {
						// characters like /\n/ can just be kept as "\n" instead of being escaped to "\\n"
						if (/[nrt0/]/.test(g1)) {
							return m;
						}
						if ('\\' === g1) {
							return '\\\\\\\\'; // escape using 4 backslashes
						}
						return '\\\\' + g1;
					} else {
						return '[^]';
					}
				});
				// escape single quotes
				source = source.replace(/'/g, "\\'");
				// wrap source in single quotes
				return "'" + source + "'";
			}
		);
		return toRenderedChunk(str);
	},
};
```

Mangled by prettier: 
```ts
const inlineRegexSourcePlugin: Plugin = {
	name: 'inline-regex-source',
	renderChunk (code) {
		const str = new MagicString(code);
		str.replace(
			/\/((?:[^\n\r[\\\/]|\\.|\[(?:[^\n\r\\\]]|\\.)*\])+)\/\s*\.\s*source\b/g,
			(m, source: string) => {
				// escape backslashes
				source = source.replace(/\\(.)|\[(?:\\s\\S|\\S\\s)\]/g, (m, g1: string) => {
					if (g1) {
						// characters like /\n/ can just be kept as "\n" instead of being escaped to "\\n"
						if (/[nrt0/]/.test(g1)) {
							return m;
						}
						if ('\\' === g1) {
							return '\\\\\\\\'; // escape using 4 backslashes
						}
						return '\\\\' + g1;
					}
					else {
						return '[^]';
					}
					if ('\\' === g1) {
						return '\\\\\\\\'; // escape using 4 backslashes
					}
					return '\\\\' + g1;
				} else {
					return '[^]';
				}
			});
			// escape single quotes
			source = source.replace(/'/g, "\\'");
			// wrap source in single quotes
			return "'" + source + "'";
		});
		return toRenderedChunk(str);
	},
};
```

DmitrySharabin commented:
> So this prettier run mangles `scriptsa/build.ts`. I added it to `.prettierignore` to avoid any issues

How about adding `prettier-ignore` and not excluding the file entirely? It seems to solve the issue and correctly adds a space after `renderChunk`.

```js
// prettier-ignore
const inlineRegexSourcePlugin: Plugin = {
```

Could you please check if it works not “on my machine” only? 🙂

robert-j-webb commented:
> > So this prettier run mangles `scriptsa/build.ts`. I added it to `.prettierignore` to avoid any issues
> 
> How about adding `prettier-ignore` and not excluding the file entirely? It seems to solve the issue and correctly adds a space after `renderChunk`.
> 
> ```js
> // prettier-ignore
> const inlineRegexSourcePlugin: Plugin = {
> ```
> 
> Could you please check if it works not “on my machine” only? 🙂

I did one better and applied it just to the lambda that gets mangled : )

DmitrySharabin commented:
I tried another approach that might be slightly better. I noticed that none of the plugins mangle the files by themselves. It only happens when we expect them to be applied sequentially. I already knew plugin order matters, and the one we use here plays well in our other project. But it seems we should swap the `prettier-plugin-space-before-function-paren` and `prettier-plugin-brace-style` here.

```json
{
	"plugins": [
		"prettier-plugin-space-before-function-paren",
		"prettier-plugin-brace-style",
		"prettier-plugin-merge"
	],
	...
}
```

Could you please check it?

DmitrySharabin commented:
Actually, it is weird that we have to do that. It's more like we are working around a bug in our plugin (which adds a space before the function paren). We need to fix the plugin first. I’ll work on it.

DmitrySharabin commented:
> Actually, it is weird that we have to do that. It's more like we are working around a bug in our plugin (which adds a space before the function paren). We need to fix the plugin first. I’ll work on it.

It turned out it's not an issue with the `prettier-plugin-space-before-function-paren` plugin (phew 😅). It's the `prettier-plugin-merge` that screws things up (in some cases). The workaround I proposed seems still valid. However, in cases where previously we had mangled blocks of code, `else` might not be on a separate line.

I have another issue to file concerning the merge plugin. It might be when that issue is fixed, we won't need any workarounds. 🤞

**UPDATE**: I hope to update the `prettier-plugin-space-before-function-paren` plugin soon, so it covers all the cases Prism codebase needs. I'll let you know when it's published.

**UPDATE 2**: Another option would be adding `"parser": "babel-ts"` to `.prettierrc` (and don't swap the plugins), but it's not ideal either—the same problem with not handled `else`.

DmitrySharabin commented:
> I'll let you know when it's published

It's published: https://www.npmjs.com/package/prettier-plugin-space-before-function-paren

robert-j-webb commented:
> > I'll let you know when it's published
> 
> It's published: https://www.npmjs.com/package/prettier-plugin-space-before-function-paren

Unfortunately still seeing the same issue. I did check `package-json.lock` to make sure I had 0.0.7

robert-j-webb commented:
Correct that changing plugin order does fix it though!

DmitrySharabin commented:
> > > I'll let you know when it's published
> > 
> > 
> > It's published: https://www.npmjs.com/package/prettier-plugin-space-before-function-paren
> 
> Unfortunately still seeing the same issue. I did check `package-json.lock` to make sure I had 0.0.7

Yeah, it is more for other issues that weren't previously addressed. I filed an [issue](https://github.com/ony3000/prettier-plugin-merge/issues/34) for the one that causes files to be mangled. I hope they can give me a hint of why we might face it at all.

LeaVerou commented on .babelrc:
```diff
@@ -1,18 +1,18 @@
-            }
-        ]
-    ]
+	"presets": [
```
I think we could actually skip babel and minification entirely in v2. If folks need broader browser support than what we provide or to minify, they can do their own transformations.

DmitrySharabin commented:
> > > > I'll let you know when it's published
> > > 
> > > 
> > > It's published: https://www.npmjs.com/package/prettier-plugin-space-before-function-paren
> > 
> > 
> > Unfortunately still seeing the same issue. I did check `package-json.lock` to make sure I had 0.0.7
> 
> Yeah, it is more for other issues that weren't previously addressed. I filed an [issue](https://github.com/ony3000/prettier-plugin-merge/issues/34) for the one that causes files to be mangled. I hope they can give me a hint of why we might face it at all.

Okay, the issue with the `prettier-plugin-merge` is fixed in the [recent version](https://www.npmjs.com/package/prettier-plugin-merge), and the file doesn't seem to be mangled anymore (and we probably don't need either fully ignore it or none of its parts). 🥳

Speaking of the plugin order, we still need our plugin to come _after_ the `prettier-plugin-brace-style` (in that case, all the changes made by both plugins will be preserved):

```json
"plugins": [
	"prettier-plugin-brace-style",
	"prettier-plugin-space-before-function-paren",
	"prettier-plugin-merge"
]
```

robert-j-webb commented:
> > > > > I'll let you know when it's published
> > > > 
> > > > 
> > > > It's published: https://www.npmjs.com/package/prettier-plugin-space-before-function-paren
> > > 
> > > 
> > > Unfortunately still seeing the same issue. I did check `package-json.lock` to make sure I had 0.0.7
> > 
> > 
> > Yeah, it is more for other issues that weren't previously addressed. I filed an [issue](https://github.com/ony3000/prettier-plugin-merge/issues/34) for the one that causes files to be mangled. I hope they can give me a hint of why we might face it at all.
> 
> Okay, the issue with the `prettier-plugin-merge` is fixed in the [recent version](https://www.npmjs.com/package/prettier-plugin-merge), and the file doesn't seem to be mangled anymore (and we probably don't need either fully ignore it or none of its parts). 🥳
> 
> Speaking of the plugin order, we still need our plugin to come _after_ the `prettier-plugin-brace-style` (in that case, all the changes made by both plugins will be preserved):
> 
> ```json
> "plugins": [
> 	"prettier-plugin-brace-style",
> 	"prettier-plugin-space-before-function-paren",
> 	"prettier-plugin-merge"
> ]
> ```

Ok, updated to new version, and update order. Everything is looking groovy : )

robert-j-webb committed:
Initial prettier config setup, with recommended .prettierrc (#3876)

Note that prettier sync is being imported in order to enable us to format test code during tests. It should not be used anywhere else

Updated eslint rules to work with prettier, including removing redundant rules.

Commit:
Change support matrix to be 16,18,20

Commit:
Ran prettier on all files
